File: 2024/Day24/main.py
from __future__ import annotations
from copy import deepcopy
from enum import IntEnum
import itertools

SAMPLE_DATA: bool = False
if SAMPLE_DATA:
    filename = "sample_data.txt"
else:
    filename = "data.txt"


with open(filename, 'r') as f:
    data = f.readlines()

class Wire:
    name: str
    value: bool | None

    def __init__(self, name: str, value: bool | None):
        self.value = value
        self.name = name

    def __repr__(self):
        return f'{self.name} = {self.value}'

    def __hash__(self):
        return hash(self.name)

    def __eq__(self, other: str | Wire):
        if isinstance(other, str):
            return self.name == other
        return self.name == other.name
    
    def has_value(self) -> bool:
        return self.value is not None


class GateType(IntEnum):
    AND = 0
    OR = 1
    XOR = 2

class Gate:
    gate_type: GateType
    inputs: tuple[Wire, Wire]
    output: Wire

    def __init__(self, input1: Wire, input2: Wire, output: Wire, gate_type: GateType):
        self.inputs = (input1, input2)
        self.output = output
        self.gate_type = gate_type

    @staticmethod
    def from_string(input: str, existing_wires: list[Wire]):
        inputs_and_type, output = input.split('->')
        try:
            output = existing_wires[existing_wires.index(output.strip())] 
        except ValueError:
            output = Wire(output.strip(), None)
            existing_wires.append(output)
        input1, gate_type, input2 = inputs_and_type.strip().split(' ')
        try:
            input1 = existing_wires[existing_wires.index(input1.strip())] 
        except ValueError:
            input1 = Wire(input1.strip(), None)
            existing_wires.append(input1)
        try:
            input2 = existing_wires[existing_wires.index(input2.strip())] 
        except ValueError:
            input2 = Wire(input2.strip(), None)
            existing_wires.append(input2)
        gate_type = GateType[gate_type.upper()]
        return Gate(input1, input2, output, gate_type)

    def __repr__(self):
        return f'{self.inputs[0]} {self.gate_type} {self.inputs[1]} -> {self.output}'

    def can_be_calculated(self) -> bool:
        return all(w.has_value() for w in self.inputs)
    
    def calculated(self) -> bool:
        return self.output.has_value()
    
    def calculate(self):
        match self.gate_type:
            case GateType.AND:
                self.output.value = self.inputs[0].value & self.inputs[1].value
            case GateType.OR:
                self.output.value = self.inputs[0].value | self.inputs[1].value
            case GateType.XOR:
                self.output.value = self.inputs[0].value ^ self.inputs[1].value

wires: list[Wire] = []
for v in data[0:data.index('\n')]:
    name, value = v.strip().split(':')
    wires.append(Wire(name.strip(), value.strip() == '1'))
gates: list[Gate] = []
for v in data[data.index('\n')+1:]:
    gates.append(Gate.from_string(v, wires))
z_wires = sorted([w for w in wires if w.name[0] == 'z'], key=lambda w: w.name, reverse=True)
while not all(w.has_value() for w in z_wires):
    for g in gates:
        if g.can_be_calculated() and not g.calculated():
            g.calculate()

part1 = int(''.join(map(lambda w: '1' if w.value else '0', z_wires)), 2)

highest_z = z_wires[0]

wrong: set[Wire] = set()
for gate in gates:
    if gate.output.name[0] == "z" and gate.gate_type != GateType.XOR and gate.output != highest_z:
        wrong.add(gate.output)
    if (
        gate.gate_type == GateType.XOR
        and gate.output.name[0] not in ["x", "y", "z"]
        and gate.inputs[0].name[0] not in ["x", "y", "z"]
        and gate.inputs[1].name[0] not in ["x", "y", "z"]
    ):
        wrong.add(gate.output)
    if gate.gate_type == GateType.AND and "x00" not in gate.inputs:
        for subgate in gates:
            if gate.output in subgate.inputs and subgate.gate_type != GateType.OR:
                wrong.add(gate.output)
    if gate.gate_type == GateType.XOR:
        for subgate in gates:
            if gate.output in subgate.inputs and subgate.gate_type == GateType.OR:
                wrong.add(gate.output)

part2 = ",".join(sorted(map(lambda w: w.name, wrong)))

print(f'{part1} {part2}')

File: 2024/Day24/main.cpp
#include <algorithm>
#include <charconv>
#include <fstream>
#include <ranges>
#include <string>
#include <string_view>
#include <unordered_map>
#include <set>
#include <unordered_set>
#include <vector>
#include <cstdio>
#include <numeric>
#include <print>
#include <cinttypes>
#include <utility>
#include <span>

namespace r = std::ranges;
namespace v = std::ranges::views;
using namespace std::string_view_literals;

constexpr bool SAMPLE_DATA = false;

std::vector<std::string> read_file(std::string_view filename) {
    std::ifstream file(filename.data());
    std::vector<std::string> lines;
    std::string line;
    while (std::getline(file, line)) {
        lines.push_back(line);
    }
    return lines;
}

struct Wire {
    std::string name;
    std::optional<bool> value;

    Wire(std::string_view name, std::optional<bool> value = std::nullopt) : name{name}, value{value} {}
    bool operator==(const Wire& other) const {
        return name == other.name;
    }
    bool operator==(std::string_view other_name) const {
        return name == other_name;
    }
    bool has_value() const {
        return value.has_value();
    }
};

enum class GateType {
    AND,
    OR,
    XOR
};

struct Gate {
    GateType type;
    std::pair<Wire*, Wire*> inputs;
    Wire* output;

    public:
    Gate(GateType type, Wire* input1, Wire* input2, Wire* output) : type{type}, inputs{input1, input2}, output{output} {}
    static Gate from_string(std::string_view input, std::vector<std::unique_ptr<Wire>>& existing_wires) {
        auto parts = input | v::split(' ') | v::transform([](auto&& str) {
            return std::string_view{str};
        }) | r::to<std::vector>();
        auto find_wire = [&](std::string_view name) -> Wire* {
            auto it = r::find_if(existing_wires, [&](const auto& w) {
                return *w == name;
            });
            if (it != existing_wires.end()) {
                return it->get();
            } else {
                auto new_wire = std::make_unique<Wire>(name);
                existing_wires.push_back(std::move(new_wire));
                return existing_wires.back().get();
            }
        };
        GateType type;
        if (parts[1] == "AND"sv) {
            type = GateType::AND;
        } else if (parts[1] == "OR"sv) {
            type = GateType::OR;
        } else if (parts[1] == "XOR"sv) {
            type = GateType::XOR;
        } else {
            throw std::runtime_error("Unknown gate type");
        }
        auto input1 = find_wire(parts[0]);
        auto input2 = find_wire(parts[2]);
        auto output = find_wire(parts[4]);
        return Gate{type, input1, input2, output};
    };
    bool can_be_calculated() const {
        return inputs.first->has_value() && inputs.second->has_value();
    }
    bool calculated() const {
        return output->has_value();
    }
    void calculate() {
        switch (type) {
            case GateType::AND:
                output->value = inputs.first->value.value() & inputs.second->value.value();
                break;
            case GateType::OR:
                output->value = inputs.first->value.value() | inputs.second->value.value();
                break;
            case GateType::XOR:
                output->value = inputs.first->value.value() ^ inputs.second->value.value();
                break;
        }
    }
};

template <>
struct std::hash<Wire> {
    constexpr static inline std::hash<std::string> hasher{};
    std::size_t operator()(const Wire& w) const noexcept {
        return hasher(w.name);
    }
};

char is_xyz(std::string_view v) {
    char c = v.front();
    return c == 'x' || c == 'y' || c == 'z';
}

int main() {
    auto lines = read_file(SAMPLE_DATA ? "sample_data.txt"sv : "data.txt"sv);
    std::vector<std::unique_ptr<Wire>> wires;
    std::vector<Gate> gates;
    bool parsing_wires = true;
    for (const auto& line : lines) {
        if (line.empty()) {
            parsing_wires = false;
            continue;
        }
        if (parsing_wires) {
            auto parts = line | v::split(':') | v::transform([](auto&& str) {
                return std::string_view{str};
            }) | r::to<std::vector>();
            if (parts.size() == 2) {
                wires.emplace_back(std::make_unique<Wire>(parts[0], parts[1].back() == '1'));
            } else {
                wires.emplace_back(std::make_unique<Wire>(parts[0]));
            }
        } else {
            gates.push_back(Gate::from_string(line, wires));
        }
    }
    
    auto z_wires = wires | v::transform([](const auto& w) { return w.get(); }) | v::filter([](const auto& w) {
        return w->name.front() == 'z';
    }) | r::to<std::vector>();
    r::sort(z_wires, [](const auto& a, const auto& b) {
        return a->name > b->name;
    });
    while (!r::all_of(z_wires, [](const auto& w) {
               return w->has_value();
           })) {
        for (auto& gate : gates) {
            if (gate.can_be_calculated() && !gate.calculated()) {
                gate.calculate();
            }
        }
    }
    auto part1bin = z_wires | v::transform([](const auto& w) {
        return w->value.value() ? '1' : '0';
    }) | r::to<std::string>();
    size_t part1 = 0;
    std::from_chars(part1bin.data(), part1bin.data() + part1bin.size(), part1, 2);
    Wire* highest_z = z_wires.front();
    std::unordered_set<Wire> wrong{};
    for (auto& gate : gates) {
        if (gate.output->name.front() == 'z' && gate.type != GateType::XOR && gate.output != highest_z) {
            wrong.insert(*gate.output);
        }
        if (gate.type == GateType::XOR
            && !is_xyz(gate.inputs.first->name)
            && !is_xyz(gate.inputs.second->name)
            && !is_xyz(gate.output->name)) {
            wrong.insert(*gate.output);
        }
        if (gate.type == GateType::AND && (gate.inputs.first->name != "x00" && gate.inputs.second->name != "x00")) {
            for (auto& subgate : gates) {
                if ((gate.output == subgate.inputs.first || gate.output == subgate.inputs.second) && subgate.type != GateType::OR) {
                    wrong.insert(*gate.output);
                }
            }
        }
        if (gate.type == GateType::XOR) {
            for (auto& subgate : gates) {
                if ((gate.output == subgate.inputs.first || gate.output == subgate.inputs.second) && subgate.type == GateType::OR) {
                    wrong.insert(*gate.output);
                }
            }
        }
    }
    auto wrong_names = wrong | v::transform([](const auto& w) {
        return w.name;
    }) | r::to<std::vector>();
    r::sort(wrong_names);
    auto part2 = wrong_names | v::join_with(',') | r::to<std::string>();
    std::print("{} {}\n", part1, part2);
    return 0;
}